pkg/pillar/cmd/domainmgr/domainmgr.go

```diff
@@ -2818,6 +2819,40 @@ func handlePhysicalIOAdapterListImpl(ctxArg interface{}, key string,
 					}
 					aa.AddOrUpdateIoBundle(log, vfIb)
 				}
+			} else if ib.Type == types.IoVCAN {
+				// Initialize (create and enable) Virtual CAN device
+				vcan, err := canbus.AddVCANLink(ib.Ifname)
+				if err != nil {
+					ib.Error = err.Error()
+					ib.ErrorTime = time.Now()
+					log.Errorf("Failed to add Virtual CAN interface: %s", err)
+				}
+				err = canbus.LinkSetUp(vcan)
+				if err != nil {
+					ib.Error = err.Error()
+					ib.ErrorTime = time.Now()
+					log.Errorf("Failed to enable virtual CAN interface: %s", err)
+				}
+			} else if ib.Type == types.IoCAN {
+				// Initialize physical CAN device
+				canIf, err := canbus.GetCANLink(ib.Ifname)
+				if err != nil {
+					ib.Error = err.Error()
+					ib.ErrorTime = time.Now()
+					log.Errorf("Failed to fetch CAN interface: %s", err)
+				}
+				err = canbus.SetupCAN(canIf, ib.Cbattr)
+				if err != nil {
+					ib.Error = err.Error()
+					ib.ErrorTime = time.Now()
+					log.Errorf("Failed to setup CAN interface: %s", err)
+				}
+				err = canbus.LinkSetUp(canIf)
```

Comment:
What happens if `ib.Ifname` is `shenanigans0`. Then `canbus.GetCANLink` returns an error and it is logged with `log.Errorf` (line 2842).
Then `canbus.SetupCAN` returns an error, because `canIf` is nil and the error is logged.
Then `canbus.LinkSetUp` is called with `canIf=nil` 
then calls `netlink.LinkSetup`, 
then calls `pkgHandle.LinkSetUp`,
which calls `link.Attrs()`, but `link` is canIf which would be nil in this case.

Or does `log.Errorf` call exit and whole pillar shuts down?

Reply:
I think all of the above log.Errorf  needs to be followed by the appropriate error return or continue based on the context.

Reply:
`log.Errorf()` is not falal, so it doesn't shuts pillar down. There would be no problem to call `SetupCAN()` with canIf nil, an error would be thrown and caught by the if. But most probably the older error would be overwritten, so I've improved the code by creating two auxiliary functions to do the CAN logic and return at the first error occurrence....

Reply:
It is not just overwriting the error, it is crashing;
you can easily try it out with a test (domainmgr_test.go):
```
func init() {
       logger, log = agentlog.Init(agentName)
}

func FuzzSetupCAN(f *testing.F) {
       f.Fuzz(func(t *testing.T,
               ifname string,
               key1 string,
               val1 string,
               key2 string,
               val2 string,
               key3 string,
               val3 string,
       ) {
               cbattrs := map[string]string{
                       key1: val1,
                       key2: val2,
                       key3: val3,
               }
               ioBundle := types.IoBundle{
                       Ifname: ifname,
                       Cbattr: cbattrs,
               }

               setupCAN(&ioBundle)
       })
}
```
if you remove the `return err`s in `setupCAN` then you see it crashing.

Btw. Did your opinion about tests now change? ;-)

Resolved!

Reply:
Or use log.Fatalf if these failures should never happen. But that is dangerous because if they happen you're likely to get into a reboot loop.